File: src/main/java/io/github/rcarlosdasilva/cogen/core/handler/CodeGenerateHandler.java
package io.github.rcarlosdasilva.cogen.core.handler;

import com.baomidou.mybatisplus.annotations.TableField;
import com.baomidou.mybatisplus.annotations.TableName;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import io.github.rcarlosdasilva.cogen.config.Configuration;
import io.github.rcarlosdasilva.cogen.config.convention.JavaType;
import io.github.rcarlosdasilva.cogen.config.various.BasicPackage;
import io.github.rcarlosdasilva.cogen.config.various.Database;
import io.github.rcarlosdasilva.cogen.core.Handler;
import io.github.rcarlosdasilva.cogen.core.Renderer;
import io.github.rcarlosdasilva.cogen.core.Storage;
import io.github.rcarlosdasilva.cogen.model.ClassDetail;
import io.github.rcarlosdasilva.cogen.model.FieldDetail;
import io.github.rcarlosdasilva.cogen.model.TableDetail;
import io.github.rcarlosdasilva.cogen.utils.Files;
import io.github.rcarlosdasilva.cogen.utils.LogIf;
import io.github.rcarlosdasilva.cogen.utils.Utils;
import lombok.extern.slf4j.Slf4j;
import org.apache.velocity.app.VelocityEngine;

import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * @author <a href="mailto:rcarlosdasilva@qq.com">Dean Zhao</a>
 */
@Slf4j
public class CodeGenerateHandler implements Handler {

  private VelocityEngine engine;
  private Configuration configuration;
  private Database database;
  List<ClassDetail> entities = Lists.newArrayList();

  @Override
  public void handle() {
    log.info("[code] - 繁殖ing");

    configuration = Storage.getConfiguration();
    database = configuration.getDatabase();

    genEntities();

    genPackages();
  }

  private void genEntities() {
    if (database == null) {
      configuration.getEntities().forEach(cls -> {
        ClassDetail cd = new ClassDetail();
        cd.setName(cls.getSimpleName());
        cd.setPck(cls.getPackage().getName());

        if (cls.isAnnotationPresent(TableName.class)) {
          TableName tn = (TableName) cls.getDeclaredAnnotation(TableName.class);
          TableDetail td = new TableDetail();
          td.setName(tn.value());

          List<FieldDetail> fields = Lists.newArrayList();
          Stream.of(cls.getDeclaredFields()).forEach(clsf -> {
            FieldDetail fd = new FieldDetail();
            if (clsf.isAnnotationPresent(TableField.class)) {
              TableField tf = clsf.getDeclaredAnnotation(TableField.class);
              fd.setName(tf.value());
            } else {
              fd.setName(clsf.getName());
            }
            fd.setBaseName(clsf.getName());
            fd.setJavaType(JavaType.of(clsf.getType().getSimpleName()));
            fields.add(fd);
          });
          td.setFields(fields);
          cd.setTable(td);
        }

        entities.add(cd);
      });
      return;
    }

    List<TableDetail> tables = Storage.getFilteredTables();
    LogIf.error(log, tables == null, "没有可以处理的表");
    tables.forEach(table -> {
      BasicPackage pck = configuration.getEntityPackage();
      String className = Utils.name(table.getBaseName(), pck.getBasicClass().getPrefix(),
          pck.getBasicClass().getSuffix(), true);

      List<String> imports = table.getFields().stream()
          .filter(filter -> filter.getJavaType() != null && filter.getJavaType().getPackageName().contains("."))
          .map(filter -> filter.getJavaType().getPackageName())
          .distinct().collect(Collectors.toList());

      ClassDetail classDetail = new ClassDetail();
      classDetail.setModule(pck.getModule());
      classDetail.setName(className);
      classDetail.setPck(pck.getName());
      classDetail.setImports(imports);
      classDetail.setTable(table);

      Map<String, Object> data = basicData();
      data.put("cls", classDetail);

      output(pck, className, data);

      entities.add(classDetail);
    });
  }

  private void genPackages() {
    configuration.getPackages().forEach(this::genOnePackage);
  }

  private void genOnePackage(BasicPackage pck) {
    entities.forEach(entity -> {
      String className = Utils.name(entity.getName(), pck.getBasicClass().getPrefix(),
          pck.getBasicClass().getSuffix(), true);

      ClassDetail classDetail = new ClassDetail();
      classDetail.setModule(pck.getModule());
      classDetail.setName(className);
      classDetail.setPck(pck.getName());

      Map<String, Object> data = basicData();
      data.put("entity", entity);
      data.put("cls", classDetail);

      output(pck, className, data);
    });
  }

  private void output(BasicPackage pck, String className, Map<String, Object> data) {
    String content = Renderer.render(pck.getTemplate(), data);

    String path = Files.path(configuration.getOut(), pck.getModulePath(), configuration.mavenPath(), pck.getName(),
        className) + pck.getBasicClass().getExtension();
    Files.write(content, path);
  }

  private Map<String, Object> basicData() {
    Map<String, Object> data = Maps.newHashMap();
    data.put("author", configuration.getAuthor());
    if (configuration.isShowTime()) {
      data.put("time", new Date());
    }
    data.put("version", configuration.getVersion());
    data.putAll(configuration.getExtra());
    return data;
  }

}

File: src/main/kotlin/io/github/rcarlosdasilva/cogen/core/handler/CodeGenerateHandler.kt
package io.github.rcarlosdasilva.cogen.core.handler

import com.baomidou.mybatisplus.annotations.TableField
import com.baomidou.mybatisplus.annotations.TableName
import com.google.common.collect.Lists
import com.google.common.collect.Maps
import io.github.rcarlosdasilva.cogen.config.BasicPackage
import io.github.rcarlosdasilva.cogen.config.JavaType
import io.github.rcarlosdasilva.cogen.core.Handler
import io.github.rcarlosdasilva.cogen.core.Renderer
import io.github.rcarlosdasilva.cogen.core.Storage
import io.github.rcarlosdasilva.cogen.model.ClassModel
import io.github.rcarlosdasilva.cogen.model.FieldModel
import io.github.rcarlosdasilva.cogen.model.TableModel
import io.github.rcarlosdasilva.cogen.util.Files
import io.github.rcarlosdasilva.cogen.util.LogIf
import io.github.rcarlosdasilva.cogen.util.name
import mu.KotlinLogging
import java.util.*

object CodeGenerateHandler : Handler {

  private val logger = KotlinLogging.logger {}

  private lateinit var entities: List<ClassModel>

  override fun handle() {
    logger.info("[code] - 繁殖ing")

    if (Storage.configuration!!.database == null) {
      scanEntities()
    } else {
      genEntities()
    }
    genPackages()
  }

  private fun scanEntities() {
    val models = mutableListOf<ClassModel>()
    Storage.configuration!!.entities.forEach { etc ->
      if (!etc.isAnnotationPresent(TableName::class.java)) {
        return@forEach
      }

      val clsModel = ClassModel(etc.simpleName)
      clsModel.pck = etc.`package`.name

      val tn = etc.getDeclaredAnnotation(TableName::class.java) as TableName
      val tabModel = TableModel(tn.value)

      val fields = Lists.newArrayList<FieldModel>()
      etc.declaredFields.forEach { etf ->
        val fldModel = if (etf.isAnnotationPresent(TableField::class.java)) {
          val tf = etf.getDeclaredAnnotation(TableField::class.java)
          FieldModel(tf.value)
        } else {
          FieldModel(etf.name)
        }
        fldModel.javaName = etf.name
        fldModel.javaType = JavaType.of(etf.type.simpleName)
        fields.add(fldModel)
      }
      tabModel.fields = fields
      clsModel.table = tabModel

      models.add(clsModel)
    }
    entities = models
  }

  private fun genEntities() {
    LogIf.error(logger, Storage.filteredTables == null, "[code] - 没有可以处理的表")

    val models = mutableListOf<ClassModel>()
    val pck = Storage.configuration!!.entityPackage!!
    Storage.filteredTables?.forEach { tabModel ->
      val className = name(tabModel.javaName!!, pck.basicClass.prefix, pck.basicClass.suffix, true)
      val imports = mutableListOf<String>()

      val clsModel = ClassModel(className)
      clsModel.module = pck.module
      clsModel.pck = pck.name
      clsModel.table = tabModel
      pck.basicClass.supers.run {
        val fieldNames = tabModel.fields!!.map { it.name }.toList()
        var sc: String? = null
        var defaultSc: String? = null
        for ((k, v) in this) {
          if (fieldNames.contains(k)) {
            sc = v
            break
          }
          if (k.isBlank()) {
            defaultSc = v
          }
        }
        clsModel.superClass = if (sc == null) {
          defaultSc
        } else {
          sc
        }
        if (clsModel.superClass != null) {
          imports.add(clsModel.superClass!!)
          clsModel.superClass = clsModel.superClass!!.split(".").last()
        }
      }

      tabModel.fields = tabModel.fields!!.filter { !it.isIgnore }
      imports.addAll(tabModel.fields?.filter {
        it.javaType != null && it.javaType?.fullName!!.contains(".")
      }?.map { it.javaType?.fullName!! }?.distinct()?.toMutableList() ?: mutableListOf())
      clsModel.imports = imports

      val data = basicData()
      data["cls"] = clsModel

      output(pck, className, data)

      models.add(clsModel)
    }
    entities = models
  }

  private fun genPackages() {
    Storage.configuration!!.packages.forEach { this.genOnePackage(it) }
  }

  private fun genOnePackage(pck: BasicPackage) {
    entities.forEach { entity ->
      val className = name(entity.name, pck.basicClass.prefix, pck.basicClass.suffix, true)

      val clsModel = ClassModel(className)
      clsModel.module = pck.module
      clsModel.pck = pck.name

      val data = basicData()
      data["entity"] = entity
      data["cls"] = clsModel

      output(pck, className, data)
    }
  }

  private fun output(pck: BasicPackage, className: String, data: Map<String, Any>) {
    val content = Renderer.render(pck.template, data)

    val path = Files.path(
      Storage.configuration!!.out!!,
      pck.module!!,
      Storage.configuration!!.mavenPath,
      pck.name,
      className
    ) + (pck.basicClass.extension ?: Storage.configuration!!.lang.extension)
    Files.write(content, path, Storage.configuration!!.isCoverage)
  }

  private fun basicData(): MutableMap<String, Any> {
    val data = Maps.newHashMap<String, Any>()
    with(Storage.configuration!!) {
      data["authorName"] = authorName
      data["authorEmail"] = authorEmail
      if (isShowTime) {
        data["time"] = Date()
      }
      if (version?.isNotBlank() == true) {
        data["version"] = version
      }
      data.putAll(extras)
    }
    return data
  }
}